File: 随想录/回溯算法/分割回文串.go
package 回溯算法

func partition(s string) [][]string {
	var ans [][]string
	var split []string
	backTracking7(0, s, split, &ans)
	return ans
}

func backTracking7(start int, s string, split []string, ans *[][]string) {
	if start == len(s) {
		temp := make([]string, len(split))
		copy(temp, split)
		*ans = append(*ans, temp)
		return
	}

	for i := start + 1; i <= len(s); i++ {
		substr := s[start:i]
		if isHuiWen(substr) {
			split = append(split, substr)
			backTracking7(i, s, split, ans)
			split = split[:len(split)-1]
		}
	}
}

func isHuiWen(s string) bool {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		if s[i] != s[j] {
			return false
		}
	}
	return true
}

File: 算法/回溯/分割回文串.py
from typing import List


class Solution:
    def partition(self, s: str) -> List[List[str]]:
        ans = []
        path = []
        self.dsf(1, s, path, ans)
        if self.isHuiwen(s):
            ans.append([s])
        return ans

    def dsf(self, start, s, path: List[int], ans):
        if len(path) > 0:
            letters = []
            cur = 0
            for i in path:
                letters.append(s[cur:i])
                cur = i
            letters.append(s[cur:])
            if all(self.isHuiwen(letter) for letter in letters):
                ans.append(letters)

        for i in range(start, len(s)):
            path.append(i)
            self.dsf(i + 1, s, path, ans)
            path.pop()

    def isHuiwen(self, letter):
        i, j = 0, len(letter) - 1
        while i < j:
            if letter[i] != letter[j]:
                return False
            i += 1
            j -= 1
        return True

if __name__ == "__main__":
    solution = Solution()
    result = solution.partition("cdd")
    print((result))
